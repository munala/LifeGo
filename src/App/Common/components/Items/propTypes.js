import PropTypes from 'prop-types';

export default {
  bucketlist: PropTypes.shape({
    name: PropTypes.string,
    id: PropTypes.string,
    createdAt: PropTypes.string,
    updatedAt: PropTypes.string,
    description: PropTypes.string,
    items: PropTypes.arrayOf(PropTypes.shape({
      name: PropTypes.string,
      id: PropTypes.string,
      createdAt: PropTypes.string,
      updatedAt: PropTypes.string,
      done: PropTypes.bool,
    })),
    comments: PropTypes.arrayOf(PropTypes.shape({
      content: PropTypes.string,
      id: PropTypes.string,
      createdAt: PropTypes.string,
      updatedAt: PropTypes.string,
    })),
  }).isRequired,
  profile: PropTypes.shape({
    id: PropTypes.string,
    username: PropTypes.string,
    displayName: PropTypes.string,
    email: PropTypes.string,
    pictureUrl: PropTypes.string,
    friends: PropTypes.arrayOf(PropTypes.shape({})),
    searchUsers: PropTypes.arrayOf(PropTypes.shape({})),
  }).isRequired,
};
