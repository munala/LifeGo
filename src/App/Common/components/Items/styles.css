.items-container {
  display: flex;
  flex-direction: column-reverse;
  justify-content: center;
  align-items: stretch;
  font-family: sans-serif;
  font-size: 14px;
}

.item-container {
  flex-direction: row;
  justify-content: space-between;
  align-items: center;
  display: flex;
  background-color: #f7f7f7;
  border-radius: 10px;
  overflow: hidden;
  margin: 5px;
  margin-top:0px;
  padding-left: 10px;
}

.item-name {
  font-size: 14px;
  display: flex;
  color: #666;
}

.items-navigator {
  display: flex;
  flex-direction: row;
  align-items: center;
  justify-content: space-between;
}

.items-nav-buttons {
  display: flex;
  margin: 5px;
  padding-left: 5px;
  flex-direction: row;
  align-items: center;
  justify-content: flex-start;
}

a {
  text-decoration: none;
  color: #777;
  margin-right: 5px;
  margin-left: 5px;
}

a:hover {
  text-decoration: underline;
}
