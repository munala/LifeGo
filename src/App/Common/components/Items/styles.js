const ITEM_HEIGHT = 48;

export default {
  darkGrey: {
    color: '#444',
  },
  menu: {
    maxHeight: ITEM_HEIGHT * 4.5,
    paddingLeft: 10,
    paddingRight: 10,
  },
};
