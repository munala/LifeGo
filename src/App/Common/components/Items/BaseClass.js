/* eslint-disable react/no-unused-state */
import { Component } from 'react';

import propTypes from './propTypes';

class BaseClass extends Component {
  onChange = ({ target: { value } }) => {
    this.setState({
      item: {
        ...this.state.item,
        name: value,
      },
    });
  }

  onDone = (item) => {
    const newItem = { ...item, done: !item.done };
    this.props.actions.updateItem(
      this.props.bucketlist,
      newItem,
    );
  }

  onSave = async () => {
    const { actions, bucketlist } = this.props;
    const { item, editMode } = this.state;
    let response;

    if (editMode) {
      response = await actions.updateItem(bucketlist, { ...item });
    } else {
      response = await actions.saveItem(bucketlist, item);
    }

    if (!response.error) {
      // TODO
    }
  }

  navigatePage = (direction) => {
    this.setState({
      page: this.state.page + (direction === 'next' ? 1 : -1),
    });
  }

  deleteItem = () => {
    this.closeMenu();
    this.props.openDialog();
  }

  delete = () => {
    this.props.closeDialog();
    this.props.actions.deleteItem(
      this.props.bucketlist,
      this.state.selectedItem,
    );
  }

  editItem = () => {
    this.closeMenu();
    this.openForm({
      type: 'Edit',
      name: 'item',
    }, this.state.selectedItem);
  }

  cancel = () => {
    this.setState({
      item: {},
      editMode: false,
    });
  }

  openForm = (context, content = {}) => {
    // TODO
  }

  openMenu = (selectedItem) => {
    // TODO
  }

  closeMenu = () => {
    // TODO
  }
}

BaseClass.propTypes = propTypes;

export default BaseClass;
