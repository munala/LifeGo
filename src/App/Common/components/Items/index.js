import React from 'react';

import NewInput from '../NewInput';
import CheckBox from '../CheckBox';
import BaseClass from './BaseClass';
import propTypes from './propTypes';
import './styles.css';

class Items extends BaseClass {
  state = {
    page: 0,
    item: {
      name: '',
    },
    selectedItem: {
      id: '',
      content: '',
    },
    typing: false,
    submitting: false,
    editMode: false,
  }

  renderItems = bucketlist => bucketlist.items
    .slice(this.state.page * 5, (this.state.page * 5) + 5)
    .map(item => (
      <div key={item.id} className="item-container">
        <div className="item-name">{item.name}</div>
        <CheckBox
          checked={item.done}
          onChange={() => this.onDone(item)}
          disabled={this.props.bucketlist.userId !== this.props.profile.id}
        />
      </div>
    ))

  render = () => {
    const { bucketlist } = this.props;
    const {
      page, editMode, submitting, item,
    } = this.state;

    return (
      <div className="items-container">
        {
          this.props.bucketlist.userId === this.props.profile.id && this.props.mode &&
          <NewInput
            content={{
              ...item,
              content: item.name,
              type: 'name',
            }}
            onChange={this.onChange}
            save={this.onSave}
            submitting={submitting}
            buttonLabel={editMode ? 'save' : 'add'}
          />
        }

        {this.renderItems(bucketlist)}

        <div className="items-nav-buttons">
          {
            bucketlist.items.length > 0 && page > 0 &&
            <a href="#" onClick={() => this.navigatePage('previous')}>previous</a>
          }
          {
            bucketlist.items.length > 0 && bucketlist.items.length > ((page + 1) * 5) &&
            <a href="#" onClick={() => this.navigatePage('next')}>next</a>
          }
        </div>
      </div>
    );
  }
}

Items.propTypes = propTypes;

Items.defaultProps = {
  bucket: {
    id: '',
    comments: [],
    items: [],
  },
  mode: null,
};

export default Items;
