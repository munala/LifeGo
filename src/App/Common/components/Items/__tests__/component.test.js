import React from 'react';
import { shallow } from 'enzyme';
import Items from '../';

describe('Items tests', () => {
  let wrapper;
  const saveItem = jest.fn().mockReturnValue({});
  const deleteItem = jest.fn().mockReturnValue({});
  const updateItem = jest.fn().mockReturnValue({});
  const bucketlist = {
    id: 1,
    items: [{ id: 1, name: 'item', done: true }],
    userId: 1,
  };

  beforeAll(() => {
    wrapper = shallow(<Items
      bucketlist={bucketlist}
      profile={{ id: 1 }}
      actions={{
        saveItem,
        deleteItem,
        updateItem,
      }}
      mode="single"
    />);
  });

  it('renders without crashing', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('handles text change', () => {
    const input = wrapper.find('NewInput');
    input.props().onChange({ target: { value: 'value' } });
    expect(wrapper.instance().state.item.name).toEqual('value');
  });

  it('handles submitting', () => {
    const input = wrapper.find('NewInput');
    input.props().onChange({ target: { value: 'value' } });

    input.props().save();

    expect(saveItem).toHaveBeenCalledTimes(1);
    expect(saveItem).toHaveBeenCalledWith(bucketlist, { name: 'value' });
  });

  it('handles marking as done', () => {
    const item = wrapper.find('CheckBox');
    item.props().onChange();

    expect(updateItem).toHaveBeenCalledTimes(1);
  });
});
