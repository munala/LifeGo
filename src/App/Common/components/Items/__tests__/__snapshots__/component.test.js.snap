// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Items tests renders without crashing 1`] = `
<div
  className="items-container"
>
  <NewInput
    buttonLabel="add"
    cancel={[Function]}
    content={
      Object {
        "content": "",
        "name": "",
        "type": "name",
      }
    }
    editMode={false}
    focus={[Function]}
    onChange={[Function]}
    pictureUrl={null}
    save={[Function]}
    submitting={false}
  />
  <div
    className="item-container"
    key="1"
  >
    <div
      className="item-name"
    >
      item
    </div>
    <CheckBox
      checked={true}
      disabled={false}
      onChange={[Function]}
    />
  </div>
  <div
    className="items-nav-buttons"
  />
</div>
`;
